drivers/adc/adc_xmc4xxx.c

```diff
@@ -0,0 +1,350 @@
+	if (ret < 0) {
+		LOG_ERR("Invalid sequence buffer size");
+		return ret;
+	}
+
+	adc_group->ICLASS[0] &= ~VADC_G_ICLASS_CMS_Msk;
+	if (resolution == 8) {
+		adc_group->ICLASS[0] |=  2 << VADC_G_ICLASS_CMS_Pos;
+	} else if (resolution == 10) {
+		adc_group->ICLASS[0] |=  1 << VADC_G_ICLASS_CMS_Pos;
+	} else if (resolution == 12) {
+		adc_group->ICLASS[0] |=  0 << VADC_G_ICLASS_CMS_Pos;
+	} else {
+		LOG_ERR("Invalid resolution");
+		return -EINVAL;
+	}
+
+	data->channel_mask = requested_channels;
+	data->buffer = sequence->buffer;
+	adc_context_start_read(&data->ctx, sequence);
+
+	return adc_context_wait_for_completion(&data->ctx);
+}
+
+static int adc_xmc4xxx_read(const struct device *dev,
+			  const struct adc_sequence *sequence)
+{
+	int ret;
+	struct adc_xmc4xxx_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, false, NULL);
+	ret = start_read(dev, sequence);
+	adc_context_release(&data->ctx, ret);
+	return ret;
+}
+
+#ifdef CONFIG_ADC_ASYNC
+static int adc_xmc4xxx_read_async(const struct device *dev,
+			       const struct adc_sequence *sequence,
+			       struct k_poll_signal *async)
+{
+	int ret;
+	struct adc_xmc4xxx_data *data = dev->data;
+
+	adc_context_lock(&data->ctx, true, async);
+	ret = start_read(dev, sequence);
+	adc_context_release(&data->ctx, ret);
+
+	return ret;
+}
+#endif
+
+static int adc_xmc4xxx_channel_setup(const struct device *dev,
+				   const struct adc_channel_cfg *channel_cfg)
+{
+	const struct adc_xmc4xxx_cfg *config = dev->config;
+	VADC_G_TypeDef *adc_group = config->base;
+	uint32_t ch_num = channel_cfg->channel_id;
+
+	if (ch_num >= XMC4XXX_CHANNEL_COUNT) {
+		LOG_ERR("Channel %d is not valid", ch_num);
+		return -EINVAL;
+	}
+
+	if (channel_cfg->differential) {
+		LOG_ERR("Differential channels are not supported");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->gain != ADC_GAIN_1) {
+		LOG_ERR("Invalid channel gain");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->reference != ADC_REF_INTERNAL) {
+		LOG_ERR("Invalid channel reference");
+		return -EINVAL;
+	}
+
+	if (channel_cfg->acquisition_time != ADC_ACQ_TIME_DEFAULT) {
+		LOG_ERR("Invalid acquisition time");
+		return -EINVAL;
+	}
+
+	/* check that the group global calibration has successfully finished */
+	if (adc_group->ARBCFG & VADC_G_ARBCFG_CAL_Msk) {
+		LOG_WRN("Group calibration hasn't completed yet");
+		return -EBUSY;
+	}
+
+	/* assign channel to a priority group */
+	adc_group->CHASS |= BIT(ch_num);
+
+	/* store the contents in the ch_num result register. This simplifies keeping track of */
+	/* results */
+	adc_group->CHCTR[ch_num] |= ch_num << VADC_G_CHCTR_RESREG_Pos;
+	/* right align the results */
+	adc_group->CHCTR[ch_num] |= BIT(VADC_G_CHCTR_RESPOS_Pos);
+
+	/* results register configuration */
+	adc_group->RCR[ch_num] = 0;
+
+	/* Add channel to the scan sequence */
+	adc_group->ASSEL |= BIT(ch_num);
+
+	return 0;
+}
+
+#define VADC_IRQ_MIN  18
+#define IRQS_PER_VADC_GROUP 4
+
+static int adc_xmc4xxx_init(const struct device *dev)
```

Comment:
This function has a mix of XMCLib code (great!) and Zephyr specific, low-level, hardware code. Is there good reason not to use the XMCLib for all low-level hardware manipulation/configuration?

Reply:
@ifyall, thank you for reviewing!
For XMCLib vs setting registers directly, it's on a case by case basis. I sometimes prefer to set registers directly to be clearer about the intention.  

Sometimes the XMCLib functions do other things, so you have to be very careful about setting the config structs. For example, look at the new version where I use XMC_VADC_GROUP_ChannelInit(). Now if I didn't set channel_config.alias_channel = -1, it would have silently introduced a bug.

Also the config structs will in general need to be zero initialized (minor point but still uses CPU cycles).